### ISSUE TITLE
Wrong line is highlighted as a bug!

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/roslyn#33995]
**Version Used**: 
VS.NET 2017 and VS.NET 2019 preview 4

**Steps to Reproduce**:

1. Create a C# console app (.net core)
2. write this code in Main:
```C#
            object o = 1;
            var a = (int)o; // OK`
            var c = (long?)o;

            Console.WriteLine("test");
```
3. run the app

**Expected Behavior**:
A bug in the line `var c = (long?)o;`

**Actual Behavior**:
The bug is reported at the line  `Console.WriteLine("test");`

![15](https://user-images.githubusercontent.com/48354902/54076805-2f826080-42b8-11e9-82e8-13eabc185d50.jpg)

Note that this doesn't happen if you change the line `var c = (long?)o;` to `var c = (long)o;`. It seems some how related to the nullable `long?`

![16](https://user-images.githubusercontent.com/48354902/54076827-62c4ef80-42b8-11e9-9713-d3387eea8c0b.jpg)

**Comment by sharwell** [comment]
@tmat I moved this to debugging first

**Comment by tmat** [comment]
@jcouv Likely caused by a misplaced sequence point.

**Comment by jcouv** [comment]
@tmat the sequence points look correct as far as I can tell:

```C#
        [Fact]
        public void Test()
        {
            const string Source = @"
class Program
{
    static void Main()
    {
        object o = 1;
        var a = (int)o; // OK`
        var c = (long?)o;

        System.Console.WriteLine(""test"");
    }
}";
            var comp = CreateCompilation(Source, options: TestOptions.DebugExe);
            CompileAndVerify(comp).VerifyIL("Program.Main", sequencePoints: "Program.Main", source: Source, expectedIL: @"
{
  // Code size       34 (0x22)
  .maxstack  1
  .locals init (object V_0, //o
                int V_1, //a
                long? V_2) //c
  // sequence point: {
  IL_0000:  nop
  // sequence point: object o = 1;
  IL_0001:  ldc.i4.1
  IL_0002:  box        ""int""
  IL_0007:  stloc.0
  // sequence point: var a = (int)o;
  IL_0008:  ldloc.0
  IL_0009:  unbox.any  ""int""
  IL_000e:  stloc.1
  // sequence point: var c = (long?)o;
  IL_000f:  ldloc.0
  IL_0010:  unbox.any  ""long?""
  IL_0015:  stloc.2
  // sequence point: System.Console.WriteLine(""test"");
  IL_0016:  ldstr      ""test""
  IL_001b:  call       ""void System.Console.WriteLine(string)""
  IL_0020:  nop
  // sequence point: }
  IL_0021:  ret
}");
        }
```

**Comment by tmat** [comment]
@r-ramesh @chuckries Any idea why would debugger show the wrong line?

**Comment by chuckries** [comment]
@Ramkumar Ramesh<mailto:Ramkumar.Ramesh@microsoft.com> are you looking at this already? Otherwise I will take a look at it.

**Comment by chuckries** [comment]
@tmat An initial look makes me think this is an ICorDebug issue. If the cast is to `long` the VS Debugger receives `ICorDebugManagedCallback2::Exception(...)` with `nOffset` set to 15, but if the cast is to `long?` then the `Exception` callback has `nOffset` set to 22.

cc @cshung

**Comment by chuckries** [comment]
@tmat @cshung Yes this appears to be an ICorDebug issue. VS doesn't make use of the `nOffset` parameter eventually, but later we use `ICorDebugILFrame::GetIP(...)` to retrieve the the IL Offset of the exception frame. This returns us the same value as the `nOffset` parameter, 15 for `(long)o` and 22 for `(long?)o`.

**Comment by cshung** [comment]
I am on it.

**Comment by VBAndCs** [comment]
I expect that `var c = (long?)o;` is lowered to two lines somehow somewhere and the bug occurs in the second line, which doesn't exit in the editor.

**Comment by gafter** [comment]
@cshung This bug demonstrates why using the return address (the instruction *following* the instruction that throws the exception) is incorrect for identifying the location from which the exception was thrown.  Subtract one from it (i.e. in the previous IL instruction) and we should have the correct line.

**Comment by chuckries** [comment]
@cshung so are you investigating more? What needs to be done with this?

**Comment by CyrusNajmabadi** [comment]
Closing out. Have conferred with @cshung that the issue has been handed off to the correct internal team.

**Comment by gafter** [comment]
This still repros. @cshung Can you please link me to the internal issue tracking this?

**Comment by tommcdon** [comment]
Runtime issue tracking this issue: https://github.com/dotnet/runtime/issues/34226
